### PR TITLE
feat(sonarr): download client

### PR DESCRIPTION
**Pull request by Fuochi**
- add downloadclient
- add downloadclientconfig
- add remotepathmapping

### CODE REVIEW

File: sonarr/downloadclient.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	return s.GetDownloadClientsContext(context.Background())
+}
+
+// GetDownloadClientsContext returns all configured downloadclients.
+func (s *Sonarr) GetDownloadClientsContext(ctx context.Context) ([]*DownloadClientOutput, error) {
+	var output []*DownloadClientOutput
+
+	req := starr.Request{URI: bpDownloadClient}
+	if err := s.GetInto(ctx, req, &output); err != nil {
+		return nil, fmt.Errorf("api.Get%s): %w", &req, err)
```
Comment (davidnewhall) [inline sonarr/downloadclient.go:59]: paren, and a few more follow

File: sonarr/downloadclient.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	return output, nil
+}
+
+// GetDownloadClient returns a single downloadclient.
```
Comment (davidnewhall) [inline sonarr/downloadclient.go:65]: "download client."

File: sonarr/downloadclient.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	client *DownloadClientInput,
+) (*DownloadClientOutput, error) {
+	var output DownloadClientOutput
+
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(client); err != nil {
+		return nil, fmt.Errorf("json.Marshal(%s): %w", bpDownloadClient, err)
+	}
+
+	req := starr.Request{URI: path.Join(bpDownloadClient, fmt.Sprint(int(client.ID))), Body: &body}
```
Comment (davidnewhall) [inline sonarr/downloadclient.go:122]: `int()` not needed.

File: sonarr/downloadclientconfig.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	config *DownloadClientConfig,
+) (*DownloadClientConfig, error) {
+	var output DownloadClientConfig
+
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(config); err != nil {
+		return nil, fmt.Errorf("json.Marshal(%s): %w", bpDownloadClientConfig, err)
+	}
+
+	req := starr.Request{URI: path.Join(bpDownloadClientConfig, fmt.Sprint(int(config.ID))), Body: &body}
```
Comment (davidnewhall) [inline sonarr/downloadclientconfig.go:57]: `int()` not needed.

File: sonarr/downloadclientconfig_test.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    "id": 1
+}`
+
+func TestGetDownloadClientConfig(t *testing.T) {
+	t.Parallel()
+
+	tests := []*starr.TestMockData{
+		{
+			Name:            "200",
+			ExpectedPath:    path.Join("/", starr.API, sonarr.APIver, "config/downloadClient"),
```
Comment (davidnewhall) [inline sonarr/downloadclientconfig_test.go:25]: `"config", "downloadClient"`

File: sonarr/downloadclientconfig_test.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+}
+
+func TestUpdateDownloadClientConfig(t *testing.T) {
+	t.Parallel()
+
+	tests := []*starr.TestMockData{
+		{
+			Name:           "202",
+			ExpectedPath:   path.Join("/", starr.API, sonarr.APIver, "config/downloadClient/1"),
```
Comment (davidnewhall) [inline sonarr/downloadclientconfig_test.go:69]:
```suggestion
			ExpectedPath:   path.Join("/", starr.API, sonarr.APIver, "config", "downloadClient", "1"),
```

File: sonarr/remotepathmapping.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	mapping *RemotePathMapping,
+) (*RemotePathMapping, error) {
+	var output RemotePathMapping
+
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(mapping); err != nil {
+		return nil, fmt.Errorf("json.Marshal(%s): %w", bpRemotePathMapping, err)
+	}
+
+	req := starr.Request{URI: path.Join(bpRemotePathMapping, fmt.Sprint(int(mapping.ID))), Body: &body}
```
Comment (davidnewhall) [inline sonarr/remotepathmapping.go:98]: `int()` not needed

File: sonarr/remotepathmapping_test.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	}
+}
+
+func TestUpdateRemotePathMapping(t *testing.T) {
+	t.Parallel()
+
+	tests := []*starr.TestMockData{
+		{
+			Name:           "201",
+			ExpectedPath:   path.Join("/", starr.API, sonarr.APIver, "remotePathMapping", "2"),
```
Comment (davidnewhall) [inline sonarr/remotepathmapping_test.go:169]: 👍

**Review by davidnewhall — CHANGES_REQUESTED** [review]

File: lidarr/tag_test.go [hunk 9]
```diff
@@ -61,7 +61,7 @@ func TestGetTag(t *testing.T) {
 	tests := []*starr.TestMockData{
 		{
 			Name:           "200",
-			ExpectedPath:   path.Join("/", starr.API, lidarr.APIver, "tag/1"),
+			ExpectedPath:   path.Join("/", starr.API, lidarr.APIver, "tag", "1"),
```
Comment (davidnewhall) [inline lidarr/tag_test.go:64]: woah you didn't ave to do all that. My comments on things like this mostly nitpicks.

File: sonarr/downloadclient.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	var output DownloadClientOutput
+
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(client); err != nil {
+		return nil, fmt.Errorf("json.Marshal(%s): %w", bpDownloadClient, err)
+	}
+
+	req := starr.Request{URI: bpDownloadClient, Body: &body}
+	if err := s.PostInto(ctx, req, &output); err != nil {
+		return nil, fmt.Errorf("api.Post%s): %w", &req, err)
```
Comment (davidnewhall) [inline sonarr/downloadclient.go:100]: a few more parens missing.  x3

File: sonarr/downloadclient.go (hunk 2 above)
Comment (Fuochi) [reply sonarr/downloadclient.go]: ok, then we'll use this standard for comments. lowercase with spaces

File: sonarr/downloadclient.go (hunk 3 above)
Comment (Fuochi) [reply sonarr/downloadclient.go]: changed across all files, there was still something missing

File: sonarr/downloadclient.go (hunk 2 above)
Comment (davidnewhall) [reply sonarr/downloadclient.go]: Yes, please. You can write method or type names in comments without spaces, but like this it looks better with a space.

File: sonarr/downloadclientconfig_test.go (hunk 6 above)
Comment (Fuochi) [reply sonarr/downloadclientconfig_test.go]: changed across all files

File: sonarr/remotepathmapping.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+	req := starr.Request{URI: bpRemotePathMapping}
+	if err := s.GetInto(ctx, req, &output); err != nil {
+		return nil, fmt.Errorf("api.Get(%s): %w", &req, err)
+	}
+
+	return output, nil
+}
+
+// GetRemotePathMapping returns a single remote path mapping.
+func (s *Sonarr) GetRemotePathMapping(mappingID int) (*RemotePathMapping, error) {
```
Comment (davidnewhall) [inline sonarr/remotepathmapping.go:42]: mapping IDs are int64

File: sonarr/remotepathmapping.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	return s.GetRemotePathMappingContext(context.Background(), mappingID)
+}
+
+// GetRemotePathMappingContext returns a single remote path mapping.
+func (s *Sonarr) GetRemotePathMappingContext(ctx context.Context, mappingID int) (*RemotePathMapping, error) {
```
Comment (davidnewhall) [inline sonarr/remotepathmapping.go:47]: int64 here too

File: sonarr/remotepathmapping.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	req := starr.Request{URI: path.Join(bpRemotePathMapping, fmt.Sprint(mapping.ID)), Body: &body}
+	if err := s.PutInto(ctx, req, &output); err != nil {
+		return nil, fmt.Errorf("api.Put(%s): %w", &req, err)
+	}
+
+	return &output, nil
+}
+
+// DeleteRemotePathMapping removes a single remote path mapping.
+func (s *Sonarr) DeleteRemotePathMapping(mappingID int) error {
```
Comment (davidnewhall) [inline sonarr/remotepathmapping.go:107]: another int64

File: sonarr/downloadclient.go [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+	req := starr.Request{URI: bpDownloadClient}
+	if err := s.GetInto(ctx, req, &output); err != nil {
+		return nil, fmt.Errorf("api.Get(%s): %w", &req, err)
+	}
+
+	return output, nil
+}
+
+// GetDownloadClient returns a single download client.
+func (s *Sonarr) GetDownloadClient(downloadclientID int) (*DownloadClientOutput, error) {
```
Comment (davidnewhall) [inline sonarr/downloadclient.go:66]: download client ID is an int64

File: sonarr/downloadclient.go [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+	req := starr.Request{URI: path.Join(bpDownloadClient, fmt.Sprint(client.ID)), Body: &body}
+	if err := s.PutInto(ctx, req, &output); err != nil {
+		return nil, fmt.Errorf("api.Put%s): %w", &req, err)
+	}
+
+	return &output, nil
+}
+
+// DeleteDownloadClient removes a single download client.
+func (s *Sonarr) DeleteDownloadClient(downloadclientID int) error {
```
Comment (davidnewhall) [inline sonarr/downloadclient.go:131]: Another in64.

**Review by davidnewhall — CHANGES_REQUESTED** [review]

**Comment by Fuochi** [comment]
I'll wait for this to merge before fixing the other #64 and #65 with the same pattern

File: lidarr/tag_test.go (hunk 9 above)
Comment (Fuochi) [reply lidarr/tag_test.go:64]: I just did find and replace and it came across many files

File: sonarr/downloadclient.go (hunk 3 above)
Comment (davidnewhall) [reply sonarr/downloadclient.go]: it looks like this is still here?

File: sonarr/remotepathmapping.go (hunk 11 above)
Comment (Fuochi) [reply sonarr/remotepathmapping.go]: same thing here, few more int which were int64

**Review by davidnewhall — APPROVED** [review]

**Pull request merged**